### PR TITLE
do not check model designation if metadata file is updated

### LINKED ISSUE #69: do not check model designation if metadata file is updated

### PR DESCRIPTION
**Pull request by vidhi-mody**
closes #69 

Link to test PRs:
Valid file: https://github.com/vidhi-mody/covid19-forecast-hub/pull/13
Invalid file: https://github.com/vidhi-mody/covid19-forecast-hub/pull/14
Changed model designation to primary: https://github.com/vidhi-mody/covid19-forecast-hub/pull/15

### CODE REVIEW

File: forecast_validation/validation_logic/metadata.py [hunk 1]
```diff
@@ -68,12 +69,21 @@ def validate_metadata_contents(metadata, filepath, cache, store: dict[str, Any],
...
-        team_names = set(model.team_name for model in project.models)
-        #print(team_names)
-        if metadata['team_name'] in team_names:
-            metadata_error_output.append('METADATA ERROR: %s has more than 1 model designated as \"primary\"' % (metadata['team_abbr']))
-            is_metadata_error = True
+        for model in project.models:
+            team_models[model.team_name].append(model.name)
+            model_designation[model.name] = model.notes
+            
+        if team_models[metadata['team_name']]:
```
Comment (Serena-Wang) [inline forecast_validation/validation_logic/metadata.py:79]: Could you add more documentation of what cases you are check here? And also modify error messages to be more descriptive?

File: forecast_validation/validation_logic/metadata.py [hunk 2]
```diff
@@ -68,12 +69,21 @@ def validate_metadata_contents(metadata, filepath, cache, store: dict[str, Any],
...
+            if metadata['model_name'] not in team_models[metadata['team_name']]:
```
Comment (Serena-Wang) [inline forecast_validation/validation_logic/metadata.py:80]: For example, this is checking the case when an existing team add a new model with "primary" designation

Comment (Serena-Wang) [inline forecast_validation/validation_logic/metadata.py:80]: For example, this is checking the case when an existing team add a new model with "primary" designation

File: forecast_validation/validation_logic/metadata.py [hunk 3]
```diff
@@ -68,12 +69,21 @@ def validate_metadata_contents(metadata, filepath, cache, store: dict[str, Any],
...
+                metadata_error_output.append('METADATA ERROR: %s has more than 1 model designated as \"primary\"' % (metadata['team_abbr']))
+                is_metadata_error = True
+            else:
+                if model_designation[metadata['model_name']] == 'secondary':
```
Comment (Serena-Wang) [inline forecast_validation/validation_logic/metadata.py:84]: I'm think we should keep a list of "primary" models for each team. Our current implicit assumption is that each team should have a primary model. In this case we are checking if an existing team changes designation from "secondary" to "primary" for an existing model. Before we throw an error, we should confirm that they do have another model with "primary" designation which causes this conflict.

Comment (Serena-Wang) [inline forecast_validation/validation_logic/metadata.py:84]: I'm thinking if we should keep a list of "primary" models for each team. Our current implicit assumption is that each team should have a primary model. In this case we are checking if an existing team changes designation from "secondary" to "primary" for an existing model. Before we throw an error, we should confirm that they do have another model with "primary" designation which causes this conflict.

File: forecast_validation/validation_logic/metadata.py [hunk 4]
```diff
@@ -68,12 +69,21 @@ def validate_metadata_contents(metadata, filepath, cache, store: dict[str, Any],
...
+                    metadata_error_output.append('METADATA ERROR: %s has more than 1 model designated as \"primary\"' % (metadata['team_abbr']))
+                    is_metadata_error = True
```
Comment (Serena-Wang) [inline forecast_validation/validation_logic/metadata.py:87]: So if a new team submits a new "primary" model, it is dealt with the implicit "else" part of the if statement at line 79? Does the "else" part also takes care of the case when a team changes "primary" model designation to "secondary"? If so, could you document these in the code?

Comment (Serena-Wang) [inline forecast_validation/validation_logic/metadata.py:87]: So if a new team submits a new "primary" model, it is dealt with the implicit "else" part of the if statement at line 79?

**Review by Serena-Wang — CHANGES_REQUESTED** [review]

**Comment by elray1** [comment]
Two comments while I'm looking at this:

 1. Why are we using Zoltar to get information about metadata rather than just reading in the metadata files from the hub repository? Using Zoltar (a) seems like it introduces an extra dependency; not all hubs may have set up syncing to Zoltar, but we'd still want to be able to do this validation; (b) potentially introduces some race conditions, e.g. if multiple PRs updating model designations come in before updates are sent to Zoltar.

 2. I am wondering if some streamlined logic for this could work along these lines? This could use the existing [get_metadata_for_model](https://github.com/reichlab/covid19-forecast-hub-validations/blob/16482f327f4e313356682584e80486fb6b024b93/forecast_validation/utilities/github.py#L43) function (perhaps with modification to get metadata contents including updates from the PR?), and I think would mean we would not need to add extra data tracking structure like keeping a list of primary models.

    - for each team for which the current pull request updates at least one metadata file:
        - get the model designation for all of that team's models, according to the state of the hub repo that would be in effect if the PR were merged
        - if more than two of that team's models have primary status, throw an error

**Comment by elray1** [comment]
We took another route to closing this issue in PR #77

**Pull request closed without merging**